test/e2e/cmd/run/run.go

```diff
@@ -486,15 +481,84 @@ func (h *helper) streamTestJobOutput(streamStatus chan<- error, client *kubernet
 		defer jl.Close()
 		outputs = append(outputs, jl)
 	}
-
 	writer := io.MultiWriter(outputs...)
-	var buffer [logBufferSize]byte
-	if _, err := io.CopyBuffer(writer, stream, buffer[:]); err != nil && err != io.EOF {
-		streamStatus <- err
-		return
+
+	// The log stream may end abruptly in some environments where the network isn't reliable.
+	// Let's retry when that happens. To avoid duplicate log entries, keep track of the last
+	// Kubernetes log timestamp: we don't want to reprocess entries with a lower timestamp.
+	// If the stream drops in between several logs that share the same timestamp, the second half will be lost.
+	lastTimestamp := time.Time{}
+	for {
+		select {
+		case <-stop:
+			log.Info("Log stream stopped")
+			return
+		default:
+			log.Info("Streaming pod logs", "name", pod)
+			stream, err := getLogStream(client, pod, h.testContext.E2ENamespace)
+			if err != nil {
+				streamStatus <- err
+				continue // retry
+			}
+			defer stream.Close()
+
+			pastPreviousLogStream := false
+			scan := bufio.NewScanner(stream)
+			for scan.Scan() {
+				line := scan.Bytes()
+
+				timestamp, logLine, err := parseLog(string(line))
+				if err != nil {
+					streamStatus <- err
+					continue
+				}
+
+				// don't re-write logs that have been already written in a previous log stream attempt
+				if !pastPreviousLogStream && !timestamp.After(lastTimestamp) {
+					continue
+				}
+
+				// new log to process
+				pastPreviousLogStream = true
+				lastTimestamp = timestamp
+				if _, err := writer.Write([]byte(logLine + "\n")); err != nil {
```

Comment:
Did we use to print the timestamp as well, or am I misremembering?

Reply:
We did not, I'm adding it in https://github.com/elastic/cloud-on-k8s/pull/3389/files#diff-b48621daf1e1c944a38d2c599b1d11c4R543.

Reply:
Makes sense. I missed that.